loading.js

```diff
@@ -0,0 +1,570 @@
+// key is user ID and value is a map with key page number and value of page's html text
+var users = new Map()
+
+var finishedCount = 0;
+var maximumConcurrentThreads = 40;
+
+var abortController = new AbortController()
+
+// use these on corresponding requests to abort requests
+var globalAbortSignal = abortController.signal
+
+// Try to load the previous value from the saved cookie
+document.getElementById("userId").value = getCookie("userId");
+
+// use in failed page request handler
+async function requestPage(userID, pageNum, attemptsLeft, abortSignal) {
+	if (attemptsLeft == 0) {
+		console.error(`Too many failed attempts rerequesting page ${pageNum}`)
+		return null
+	}
+	const requestedDoc = fetch(CORSprefix + "https://rule34.xxx/index.php?page=favorites&s=view&id=" + userID + "&pid=" + ((pageNum-1)*50), {signal:abortSignal})
+		.then( async (response) => {
+			if (!response.ok) {
+				throw new Error(`Failed to request page ${pageNum}. Retrying with ${attemptsLeft-1} attempts left ...\nError ${response.status} description: ${response.statusText}`)
+			} else {
+				// when response received
+				// Get the response text and cache it
+				const responseText = await response.text();
+				cacheResponse(userID, pageNum, responseText)
+				console.log(`Successfully rerequested page ${pageNum}`)
+
+				// Parse the document
+				const parser = new DOMParser();
+				let doc = parser.parseFromString(responseText, "text/html");
+
+				return doc
+			}
+		})
+		.catch( async (error) => {
+			if (error instanceof DOMException) {
+				console.log(`Aborted request for page ${pageNum}`)
+				return null
+			}
+			console.error(error.message)
+			// Try again if request failed
+			sleep(500)
+			return requestPage(userID, pageNum, attemptsLeft-1, abortSignal)
+		})
+	const output = await requestedDoc
+	return output
+}
+
+// handles failed page requests
+function failedPageRequestHandler(userID, pageNum) {
+	const list = document.getElementById('failedRequestList')
+
+	// add button list item element to handle failed request
+	const listItem = document.createElement('li')
+	listItem.style.position = 'relative'
+	const button = document.createElement('button')
+	button.type = "button"
+	button.classList.add('rerequest')
+	button.classList.add('failed')
+	
+	const label = document.createElement('label')
+	label.style = "position: absolute; padding: 4px; margin-left: 10px"
+	label.textContent = `Failed to request page ${pageNum} for user ${userID}`
+
+	listItem.appendChild(button)
+	listItem.appendChild(label)
+
+	list.appendChild(listItem)
+
+	button.addEventListener('click', async evt => {
+		//disallow double clicks
+		if (!evt.detail || evt.detail == 1) {
+			button.disabled = true
+			// change button class to show pending status
+			button.classList.add('pending')
+			button.classList.remove('failure')
+			label.textContent = `Currently rerequesting page ${pageNum} for user ${userID}`
```

Comment:
For this and all the other similar messages, I don't think it needs to say "for user userId" since the userId should be obvious.

Reply:
Clarity is good